### DISCUSSION TITLE
How can I tell the converted model precision?

### DISCUSSION THREAD

**Discussion by shaharyuval** [openvinotoolkit/openvino#21138] [Q&A]
I'd like to convert a Pytorch model to IR representation and would like to ensure similar behavior.  I currently see a degradation in Openvino relative to Pytorch.  My first thought is to verify the model is still in FP32 precision.  How can I do so?

This is my conversion and inference code:
```
torch.onnx.export(model, torch.randn(1, 3, 244, 244), onnx_model_path)
ov_model = openvino.convert_model(onnx_model_path)
compiled_model = core.compile_model(model=ov_model, device_name='GPU')

#inference
results = compiled_model([ov_input_img])
```

Is there any attribute controlling the IR model precision?  I am unable to find documentation for this function.

I see some sources refer to another conversion function (is it older or deprecated?):
```
ov_model = openvino.tools.mo.convert_model(input_model=onnx_model_path,
                                    input_shape=[1, 3, 244, 244],
                                    layout='NCHW', ...)
```

Is it possible to control the precision with this one?  Where can I find documentation for the parameters?

**Comment by ilya-lavrenov** [comment]
Hi @shaharyuval 
`openvino.convert_model` does not change the precision of an original model. Right, `openvino.tools.mo.convert_model` is deprected one and `openvino.convert_model` is recommended to use since it's significantly simpler and has less parameters.

I see that you are using GPU, which by default performs computations in FP16. Please, try the following:
- Inference on CPU which performs FP32 computations. 
- Try to change inference precision on GPU using `core.compile_model(ov_model, 'GPU', {inference_precision: Type.f32})`

**Reply by shaharyuval** [reply]
Hi @ilya-lavrenov, 
Thank you for your reply.  
I am unable to find the ```inference_precision``` attribute.  I see some sources refer to this parameter via the namespace:
```openvino.properties.hint```, but that one doesn't appear to be installed either.  How can I access these advanced attributes?  I installed openvino using the pip packages: ```openvino``` and ```openvino-dev```.

Is there a more detailed description of all attributes supported by ```core.compile_model```.  I feel that I can only pick up bits and pieces from various questions on Stackoverflow, etc.

Thanks,

Shahar

**Reply by shaharyuval** [reply]
Hi @ilya-lavrenov,

After further tinkering with your suggested code above, I determined that this syntax is correct:
```
from openvino.runtime import Type
core.compile_model(ov_model, 'GPU', {'inference_precision': Type.f32})
```

Unfortunately, the Intel GPU does not support float32 and I get an error message.

But, when running in CPU mode I see the Pytorch and Openvino models are in agreement.  So this is clearly a float16 GPU limitation.

What are my options for tuning the model for GPU?  

Also, at present I don't see a significant performance benefit to running on GPU (Openvino on CPU and GPU run about as fast).  Should I be able to squeeze more out of the GPU?

Regards,

Shahar

**Reply by ilya-lavrenov** [reply]
Hi @shaharyuval 
What error do you observe on GPU? GPU is expected to support FP32
Regarding the performance on CPU vs GPU - it heavily depends on the devices itself. When you have powerful CPU and integrated GPU - CPU can definitely beat GPU performance.
GPU inference is not always about performance, it's also to migrate compute from CPU to leave CPU compute resources for different tasks.

**Reply by akuporos** [reply]
Hello @shaharyuval,
To use `inference_precision` please try next import:
`from openvino.properties.hint import inference_precision`

**Reply by shaharyuval** [reply]
Hi @ilya-lavrenov,
This is the error I get:

```
    compiled_model = core.compile_model(ov_model, 'GPU', {'inference_precision': Type.f32})
  File "C:\Users\...\lib\site-packages\openvino\runtime\ie_api.py", line 543, in compile_model
    super().compile_model(model, device_name, {} if config is None else config),
RuntimeError: Exception from src\inference\src\core.cpp:117:
[ GENERAL_ERROR ] Check 'supported' failed at src\plugins\intel_gpu\src\runtime\execution_config.cpp:111:
[GPU] Attepmpt to set user property inference_precision (f32) which was not registered or internal!
```

@akuporos, @ilya-lavrenov , am I supposed to invoke the call to compile_model using the ```inference_precision``` object or the string ```'inference_precision'```.  When I tried providing the key-value pair ```{inference_precision: Type.f32}```, I got the following error message:

```
    super().compile_model(model, device_name, {} if config is None else config),
TypeError: compile_model(): incompatible function arguments. The following argument types are supported:
    1. (self: openvino._pyopenvino.Core, model: openvino._pyopenvino.Model, device_name: str, properties: Dict[str, object]) -> ov::CompiledModel     
    2. (self: openvino._pyopenvino.Core, model: openvino._pyopenvino.Model, properties: Dict[str, object]) -> ov::CompiledModel
    3. (self: openvino._pyopenvino.Core, model_path: object, device_name: str, properties: Dict[str, object]) -> ov::CompiledModel
    4. (self: openvino._pyopenvino.Core, model_path: object, properties: Dict[str, object]) -> ov::CompiledModel
```

Regards,

Shahar

**Reply by akuporos** [reply]
> am I supposed to invoke the call to compile_model using the inference_precision object or the string 'inference_precision'

Since 2022.1 the recommended way is to use object from `openvino.runtime.properties`. Since 2023.2 they are available from the `openvino.properties` namespace.
Which version of OpenVINO do you use?

**Reply by shaharyuval** [reply]
Hi @akuporos,
Apparently I was using 2023.1.  After upgrading to 2023.2, I see that ```inference_precision``` appears under ```openvino.properties```. 
Also after this upgrade, ```core.compile_model(ov_model, 'GPU', {inference_precision: Type.f32})``` no longer issues an error,

However, I still see a degradation relative to the Openvino CPU model and the Pytorch model (the latter two are in agreement).  

Could there be other implicit conversions going on when moving the network to the GPU?

**Reply by akuporos** [reply]
@maxnick @vladimir-paramuzov could you please share your opinion here?

**Reply by vladimir-paramuzov** [reply]
@shaharyuval GPU plugin may do few conversions for unsupported integer types (like int64 -> int32) in such case, but with `{inference_precision: Type.f32}` set, fp16 is not supposed to be used. You can check actual execution precision using runtime model feature:
```python
runtime_model = compiled_model.get_runtime_model()
serialize(runtime_model, "exec_graph.xml")
```
and then inspect this "exec_graph.xml" as regular IR and check used precision for each layer. But for now it just looks like a bug in GPU plugin, so we would need some info from your side to fix that (IR at least)

**Reply by shaharyuval** [reply]
Hi @vladimir-paramuzov ,
I serialized the model and the xml file shows FP32 for all layers.
Regards,

Shahar
